sharding/nodesCoordinator/sovereignIndexHashedNodesCoordinator.go

```diff
@@ -0,0 +1,234 @@
+	}
+	log.Info("new nodes config is set for epoch", "epoch", arguments.Epoch)
+	currentNodesConfig := ihnc.nodesConfig[arguments.Epoch]
+	if currentNodesConfig == nil {
+		return nil, fmt.Errorf("%w epoch=%v", ErrEpochNodesConfigDoesNotExist, arguments.Epoch)
+	}
+
+	currentConfig := nodesConfig[arguments.Epoch]
+	if currentConfig == nil {
+		return nil, fmt.Errorf("%w epoch=%v", ErrEpochNodesConfigDoesNotExist, arguments.Epoch)
+	}
+
+	displaySovereignNodesConfiguration(
+		currentConfig.eligibleMap,
+		currentConfig.waitingMap,
+		currentConfig.leavingMap,
+		make(map[uint32][]Validator))
+
+	ihnc.epochStartRegistrationHandler.RegisterHandler(ihnc)
+	return ihnc, nil
+}
+
+func checkSovereignArguments(arguments ArgNodesCoordinator) error {
+	if arguments.ShardConsensusGroupSize < 1 {
+		return ErrInvalidConsensusGroupSize
+	}
+	if arguments.NbShards != 1 {
+		return ErrInvalidNumberOfShards
+	}
+
+	return checkNilArguments(arguments)
+}
+
+func (ihnc *sovereignIndexHashedNodesCoordinator) setNodesPerShards(
+	eligible map[uint32][]Validator,
+	waiting map[uint32][]Validator,
+	leaving map[uint32][]Validator,
+	epoch uint32,
+) error {
+	ihnc.mutNodesConfig.Lock()
+	defer ihnc.mutNodesConfig.Unlock()
+
+	nodesConfig, ok := ihnc.nodesConfig[epoch]
+	if !ok {
+		log.Debug("Did not find nodesConfig", "epoch", epoch)
+		nodesConfig = &epochNodesConfig{}
+	}
+
+	nodesConfig.mutNodesMaps.Lock()
+	defer nodesConfig.mutNodesMaps.Unlock()
+
+	if eligible == nil || waiting == nil {
+		return ErrNilInputNodesMap
+	}
+
+	nbNodesShard := len(eligible[core.SovereignChainShardId])
+	if nbNodesShard < ihnc.shardConsensusGroupSize {
+		return ErrSmallShardEligibleListSize
+	}
+	numTotalEligible := uint64(nbNodesShard)
+
+	err := ihnc.baseSetNodesPerShard(nodesConfig, numTotalEligible, eligible, waiting, leaving, epoch)
+	if err != nil {
+		return err
+	}
+
+	nodesConfig.nbShards = 1
+	return nil
+}
+
+// ComputeConsensusGroup will generate a list of validators based on the eligible list
+// and each eligible validator weight/chance
+func (ihnc *sovereignIndexHashedNodesCoordinator) ComputeConsensusGroup(
+	randomness []byte,
+	round uint64,
+	shardID uint32,
+	epoch uint32,
+) (validatorsGroup []Validator, err error) {
+	var selector RandomSelector
+	var eligibleList []Validator
+
+	log.Trace("computing consensus group for",
+		"epoch", epoch,
+		"shardID", shardID,
+		"randomness", randomness,
+		"round", round)
+
+	if len(randomness) == 0 {
+		return nil, ErrNilRandomness
+	}
+
+	ihnc.mutNodesConfig.RLock()
+	nodesConfig, ok := ihnc.nodesConfig[epoch]
+	if ok {
+		if shardID != core.SovereignChainShardId {
+			log.Warn("shardID is not ok, expected a sovereign chain id", "shardID", shardID, "nbShards", nodesConfig.nbShards)
+			ihnc.mutNodesConfig.RUnlock()
+			return nil, ErrInvalidShardId
+		}
+		selector = nodesConfig.selectors[shardID]
+		eligibleList = nodesConfig.eligibleMap[shardID]
+	}
+	ihnc.mutNodesConfig.RUnlock()
+
+	if !ok {
```

Comment:
L177-L179 could be moved right after L165. Doing so you can remove L166 and L174 ( if ok{} ). Also be careful to add ihnc.mutNodesConfig.RUnlock() before returning in case of !ok.